README
Dirichlet process/Gibbs sampler word segmentation program
(c) Sharon Goldwater sgwater@brown.edu, Aug. 2006
v. 1.1: 12/12/07
v. 1.2: 6/5/08
v. 1.2.1: 1/02/16
See also release_notes.txt.

(This program also contains code written by Mark Johnson and Eugene
Charniak, and minor patches by Margaret Fleck, Dan Gildea, and Antonios
Anastasopoulos.)

This program may be used freely for research purposes.  Please
acknowledge the author in any publications containing results produced
by this program or code derived from this program.

This code is made available with no guarantees of functionality,
correctness, or support.  I would appreciate knowing about any bugs
you may find, but I will not necessarily fix them in a timely fashion.
Below is the information you will need to run the program with common
options; if you wish to use other options or modify the code, you are
on your own.

----------------------------------------

Overview:

This program implements the Gibbs sampling algorithms for the
Dirichlet process and hierarchical Dirichlet process word segmentation
models described in

A Bayesian Framework for Word Segmentation: Exploring the Effects of
Context.  Sharon Goldwater, Thomas L. Griffiths, and Mark Johnson.  In
submission.

Note that the Dirichlet process model and an earlier version of the
hierarchical Dirichlet process model were originally described in

Contextual Dependencies in Unsupervised Word Segmentation.  Sharon
Goldwater, Thomas L. Griffiths, and Mark Johnson.  Proceedings of
Coling/ACL, Sydney, 2006.

and in more detail in

Nonparametric Bayesian Models of Lexical Acquisition.  Sharon
Goldwater. Ph.D. thesis, Brown University, 2006.

However, the current version of this program implements a slightly
different HDP model, uses fewer approximations during inference, and
also fixes a bug in the previous HDP implementation.

All papers are available at http://www.stanford.edu/~sgwater/

----------------------------------------

To compile:

I have successfully compiled this code using g++ 4.1.2 (linux)
	and 3.4.4 (cygwin).

make [opt | segment] : compiles optimized version (this is what
 you almost certainly want).
make dbg : compiles with -g to allow debugging
make prf : compiles to allow profiling
make nrm : compiles non-optimized version (this turns on
 lots of assertions and consistency checks, which will make the
 code extremely slow for realistic data sets).

----------------------------------------

Usage: segment [options] [input_file]

input_file: if no argument is given, the program will assume the input
file "test.in".  Input should contain a single utterance per line,
with spaces indicating correct word boundaries.  (The boundaries
provided are used for scoring only; an input file with no word
boundaries will work just fine otherwise.)

common options:
See ACL '06 paper or submitted article for more explanation.
Default hyperparameter values are
 a=20, b=.5 for the unigram model.
 a=3000, A=100, b=.2, U=.5 for the bigram model.
Other default values are given in parentheses.  Note that
the default number of iterations may be OK for preliminary
tests, but is almost certainly too low for convergence. A
better value would be 10-20k.

-h : prints usage info
-V : prints version number.
-a <alpha_0> : concentration parameter for unigram DP
-A <alpha_1> : concentration parameter for bigram HDP
-b <p_#> : prior probability of word boundary
-U <p_$> : prior probability of utterance boundary in HDP
-u t : runs the bigram model. (Otherwise runs unigram.  
	Don't use other values for -u.)
-H : turns on hyperparameter sampling, instead of using fixed
        hyperparameters.  Gives poor results.
-i <iters> : number of iterations of sampling (=1000)
-I [utt|pho|ran|true] : boundary initialization (=ran).
	none/all/random/true boundaries.
-e [samp|gmax] : type of evaluation. (= samp)
	samp returns a single sample at temperature = 1.
	gmax returns a MAP approximation by annealing
	further, down to 0.  (Note: this multiplies the
	number of iterations specified by 3.)  
	In practice, I have found little difference between these two.
-r <seed> : random number seed (if you want to reproduce results).
-w <iters> : see below
-t <iters> : see below
-o <file_prefix> : see below
-v <verbose_level> : see below

----------------------------------------

Output:

Status indicators are written to stderr (a dot is written
every ten iterations, and the "temperature" is written
when it changes.  
**** NOTE: references to "temperature" in output and
code-internally are actually 1/temperature.

Output is written to stdout.  If no options are given,
output consists of a list of all the parameter settings,
followed by summary statistics of the reference lexicon,
the segmented lexicon, the algorithm's performance
relative to the boundaries given in the input file, and
the total log probability of the proposed solution.
(See example in test.out.) Performance is listed as follows:

P 26.32 R 27.03 F 26.67 BP 51.72 BR 53.57 BF 52.63 LP 29.41 LR 26.32 LF 27.78

where the metrics are Precision, Recall, and F-score on 
tokens, boundaries (not including boundaries at utterance
edges), and types (lexicon) respectively.

output options:

-v <verbose_level> : default is 0.  -v1 will print the
	final segmentation to stdout.  -v2 will print 
	the final lexicon to stdout. -v3 prints both.
-w <N> : prints the current segmentation every N iterations
	to the file specified by the -o option (required).
	Using -w0 prints only the final segmentation.
-t <N> : prints trace statistics every N iterations
	to the file specified by the -o option (required).
-o <file_prefix> : use with -w or -t to specify output file.
	'-o file' prints to 'file.words' and/or 'file.stats'.

----------------------------------------

Examples:

%> segment my_data

(uses the input file my_data for unigram segmentation;
prints to stdout.)

%> segment -w0 -o test_unigram > test_unigram.out

(uses the input file test.in for unigram segmentation;
produces output test_unigram.out and test_unigram.words,
which are provided in this distribution as examples.)

%> segment -v3 -ut > test_bigram.out

(uses the input file test.in for bigram segmentation;
produces verbose output test_bigram.out, which is
provided in this distribution as an example.)

----------------------------------------

Additional notes:

1. References to annealing temperature in output and
code-internally are actually 1/temperature.

2. References to "monkeys" or "monkey model" in output
and code-internally indicate the unigram phoneme
model (monkeys typing at keyboards).


release_notes.txt
Version 1.2.1 -- 1/02/16

This version has some small updates contributed by Antonios
Anastasopoulos which allow the code to compile under newer versions of
g++. There is also a new flag in the Makefile (-std=c++0x) which I
needed to make these work under g++ 4.4.7. Antonios used a different
flag (-std=c++11) which did not work for me but I expect this differs
depending on your compiler and you may need to try both.

Version 1.2 -- 6/05/08

Included in this distribution:
1.  Code and example files for DP and HDP word segmentation models
    (see below and README for additional information).
2.  score_seg.prl, a script for scoring and error analysis
    (see comments at the top of the script for more information).

This version implements full table-tracking in the bigram model
rather than using the Antoniak approximation described in the
ACL 06 paper.  The HDP model is also slightly different in
its handling of utterance boundaries; full details and updated
results can be found on my web site in

A Bayesian Framework for Word Segmentation: Exploring the Effects of
Context.  Sharon Goldwater, Thomas L. Griffiths, and Mark Johnson.
Cognition 112 (1), pp. 21--54. 2009.

Note that results are quite similar to version 1.1, which fixed the
bug in the original implementation, but still used the Antoniak
approximation.  Optimal hyperparameter values are slightly different
than in version 1.1: On the Bernstein-Ratner-Brent data set, the
optimal values are approximately alpha0 = 3000, alpha1 = 100, p# = .2,
p$ = .5.  These values are now the defaults in this program, or can
be set using command-line arguments -a3000, -A100, -b.2, -U.5:

segment -ut -i20000 -a3000 -A100 -b0.2 -U.5 -Iran br-phono.txt

P 75.27 R 69.27 F 72.14 BP 90.71 BR 80.47 BF 85.28 LP 62.81 LR 53.82 LF 57.97

****************************************

Version 1.1 -- 12/12/07

This version is an interim release fixing a bug in the previous version,
but not including several planned changes (see below).

Compilation:

1.  Now compiles under (at least) g++ 3.4, 4.1.  May no longer compile
under 3.3.
2.  Typing "make" now builds the optimized version by default (still
called "segment").

Bug fixes:

I have corrected an implementation bug in which alpha0 was
accidentally used instead of alpha1 in a couple of places.  This bug
only affected the results of the bigram model. The program should now
correctly implement the bigram model described in our ACL 06 paper.
Be aware, however, that since publication we realized that the
approximation we mention for estimating the number of tables assigned
to each word in the bigram model does not take into account the fact
that some words have higher probability under the unigram generator
(P_0) than others.  Thus, it is less accurate than we had previously
believed.

Parameter values/expected results:

Optimal parameter values have changed significantly for the bigram
model.  On the data set used in our ACL paper, the optimal values are
approximately alpha0 = 3000, alpha1 = 300, p# = .2.  These values are
now the defaults in this program, or can be set using command-line
arguments -a3000, -A300, -b.2.  In general, the larger the value of
-A, the fewer boundaries will be inserted.  For a fixed value of -A,
larger values of -a will tend to produce larger lexicon sizes, and
mostly affects lexicon scores.  Here is an example result running with
the following parameter values:

segment -ut -i20000 -a3000 -A300 -b0.2 -Iran br-phono.txt

P 75.71 R 72.07 F 73.84 BP 89.88 BR 83.76 BF 86.72 LP 63.09 LR 50.34 LF 56

Note that these scores are several points lower than the best scores
in the previous (incorrect) version.  Hopefully this doesn't ruin
anyone's day (and I expect if you're trying to do better than us,
you'll be quite pleased).



segment.cc
#include <iostream>
#include <fstream>
#include <string>
#include "ECArgs.h"
#include "typedefs.h"
#include "utils.h"
#include "Utterance.h"
#include "Scoring.h"
#include "Datafile.h"
#include "State.h"

using namespace std;
// global variables
Count debug_level;
Float HYPERSAMPLING_RATIO(.1); // the standard deviation for new hyperparm proposals
bool SAMPLE_HYPERPARAMETERS(0);

int main(int argc, char* argv[])
{
  //list the options that require arguments
  ECArgs arguments(argc, argv, string("aAbUmuMiIqvreotwWT"));
  if (arguments.isset('h')) {
    cout << "Usage: segment [input_file]" << endl
	 << "-l (print reference lexicon stats without running EM)" << endl
	 << "-a <alpha0> (total unigram generator weight)" << endl
	 << "-A <alpha1> (total bigram generator weight)" << endl
	 << "-b <prior prob of word boundary>" << endl
	 << "-U <base prob of utt boundary>" << endl
	 << "-H (turn on sampling of hyperparameters)" << endl
	 << "-m [m|m2|se] (P_w: unigram generator model)" << endl
	 << "\t m: uniform monkeys" << endl
	 << "\t m2: monkeys w/ true phoneme distr." << endl
	 << "-u [m|u|y|k|b] (P_u: bigram generator model)" << endl
	 << "\t m: uniform monkeys" << endl
	 << "\t m2: monkeys w/ true phoneme distr." << endl
	 << "\t u: unigrams, sampled tokens" << endl
	 << "\t t: unigrams, sampled tables" << endl
	 << "\t y: unigrams, true types" << endl
	 << "\t k: unigrams, true tokens" << endl
	 << "\t k: bigrams, true types" << endl
	 << "\t (do not use -u if you want only unigram model)" << endl
	 << "\t M <mixing param> (proportion of noise in generator)" << endl
	 << "-i <number of iterations>" << endl
	 << "-I [utt|pho|ran|true|True] (type of init.  Default = random.)" << endl
      	 << "-e [samp|lmax|gmax] (type of evaluation)" << endl
	 << "-o <output_filename_base>" << endl
	 << "-q <Q> (print results summary every Q iters to stdout)" << endl
	 << "-t <N> (print trace statistics every N iters to output.stat)" << endl
	 << "-w <N> (print segmentation every N iters to output.words, with -w0 printing only final segmentation.)" << endl
	 << "-W <N> (print trace statistics and segmentation every N iters starting halfway through final annealing to output.stats and output.words)" << endl
	 << "-T <T> (maintain constant temperature T)" << endl
	 << "-V (prints version number)" << endl
	 << "-v N (verbose level)" << endl
	 << "\t 1: print segmentations" << endl
	 << "\t 2: print final lexicon" << endl
	 << "\t 3: print both" << endl
	 << "-d (debug)" << endl
	 << "-r random seed" << endl;
    exit(0);
  }
  if (arguments.isset('V')) {
    cout << "Version: 1.2" << endl;
    exit(0);
  }
   //input file
  string filename;
  if (arguments.nargs() > 0) {
    filename = arguments.arg(0);
  }
  else {
    filename = "test.in";
  }
  // output files and printing frequencies
  string file_base;
  ofstream words_os;
  if (arguments.isset('o')) 
    file_base = arguments.value('o').c_str();
  ofstream stats_os;
  bool print_stats = false;
  bool print_end = false;
  Count stats_freq = 0;
  if (arguments.isset('t') || arguments.isset('W')) {
    if (file_base == "") {
      cerr << "options t, w, and W require option o" << endl;
      exit(0);
    }
    string file = file_base + ".stats";
    stats_os.open(file.c_str());
    if (arguments.isset('t')) {
      print_stats = true;
      stats_freq = strtol(arguments.value('t').c_str(), NULL, 10);
      if (arguments.isset('W')) {
	cerr << "options t and W are incompatible" << endl;
	exit(0);
      }
    }
    if (arguments.isset('W')) {
      print_end = true;
      stats_freq = strtol(arguments.value('W').c_str(), NULL, 10);
    }
  }
  bool print_words = false;
  Count words_freq = 0;
  if (arguments.isset('w') || arguments.isset('W')) {
    if (file_base == "") {
      cerr << "options t, w, and W require option o" << endl;
      exit(0);
    }
    string file = file_base + ".words";
    words_os.open(file.c_str());
    if (arguments.isset('w')) {
      print_words = true;
      words_freq = strtol(arguments.value('w').c_str(), NULL, 10);
      if (arguments.isset('W')) {
	cerr << "options w and W are incompatible" << endl;
	exit(0);
      }
    }
    if (arguments.isset('W')) 
      words_freq = strtol(arguments.value('W').c_str(), NULL, 10);
  }
  // additional output stuff
  debug_level = 0;
  if (arguments.isset('d')) 
    debug_level = stringToInt(arguments.value('d'));
  int verbose_level = 0;
  if (arguments.isset('v')) {
    verbose_level = stringToInt(arguments.value('v'));
    if (verbose_level < -1 || verbose_level > 3) {
      cout << "verbose level must be -1 to 3.  Setting verbose = 0."  << endl;
      verbose_level = 0;
    }
  }
  // The stl uses lrand48() if possible, otherwise
  // rand(), so we need to seed the correct one.
  //  string seed_type;
  int seed;
  if (arguments.isset('r')) {
    seed = strtol(arguments.value('r').c_str(), NULL,10);
  } 
  else {
    seed = time(0);
  }
  srand(seed);
  try {
    DatafileBase* data = new Datafile(filename);
    Scoring scoring;
    Float alpha = 20;
    Float alpha1 = 0;
    Float p_boundary = .5;
    Float p_utt_boundary = .5;
    int ngram = 1; //word dependency model. 1=unigram. 2=bigram.
    string uni_model("m");
    if (arguments.isset('m'))
      uni_model = arguments.value('m');
    string bi_model("m");
    if (arguments.isset('u')) {
      bi_model = arguments.value('u');
      ngram = 2;
      alpha = 3000;
      alpha1 = 100;
      p_boundary = .2;
    }
    if (arguments.isset('a')) 
      alpha = strtod(arguments.value('a').c_str(), NULL);
    if (arguments.isset('A')) 
      alpha1 = strtod(arguments.value('A').c_str(), NULL);
    if (arguments.isset('b')) 
      p_boundary = strtod(arguments.value('b').c_str(), NULL);
    if (arguments.isset('U')) 
      p_utt_boundary = strtod(arguments.value('U').c_str(), NULL);
    if (arguments.isset('H')) 
      SAMPLE_HYPERPARAMETERS = true;
    string b_init("ran");
    if (arguments.isset('I'))
      b_init = arguments.value('I');
    Float noise = .0001;
    if (arguments.isset('M')) 
      noise = strtod(arguments.value('M').c_str(), NULL);

    cout << "Segmenting " << filename;
    cout << " using " << ngram << "-gram model" << endl;
    cout << "Boundary initialization: " << b_init << endl;
    Utterance::set_init(b_init);
    State::set_models(uni_model,bi_model,ngram,noise);
    State state(data, alpha, p_boundary, alpha1, p_utt_boundary);

    Count iters = 1000;
    if (arguments.isset('i'))
      iters = strtol(arguments.value('i').c_str(), NULL, 10);
    cout << "Sampling " << iters 
	    << " iterations" << endl;
    string eval = "samp";
    if (arguments.isset('e'))
      eval = arguments.value('e');
    if (eval == "lmax")
	cout << "evaluating using local max" << endl;
    else if (eval == "gmax") 
      cout << "evaluating using global max" << endl;
    else if (eval == "samp") 
      cout << "evaluating a sample" << endl;
    else
      error("unknown evaluation method");
    Count print_freq = 0;
    if (arguments.isset('q')) {
      print_freq = strtol(arguments.value('q').c_str(), NULL, 10);
      cerr << "printing status every " << print_freq << " iters" << endl;
    }
    if (verbose_level == 4) {
      cout << state << endl;
    }
    //cout << state.get_lexicon() << endl;
    cout << "random seed = " << seed << endl;
    cout << "alphabet size = " << state.alphabet_size() << endl;
    //annealing stuff
    bool anneal = true;
    Count temp_incr = 10;  //how many increments of temperature to get to T = 1
    if (iters && iters < temp_incr) temp_incr = iters;
    Count iter_incr = iters/temp_incr; //raise temp each iter_incr iters
    Float temp;
    Fs temperatures;
    if (b_init == "True" || arguments.isset('T')) {
      anneal = false;
      temp = 1;
      if (arguments.isset('T'))
	temp = strtod(arguments.value('T').c_str(), NULL);
      cout << "Not doing annealing. T = " << temp << endl;
    }
    else {
      temp = 0;
      for (Count i = 1; i <=temp_incr; i++) {
	temp += 1.0/temp_incr;
	temperatures.push_back(temp);
      }
      if (eval == "gmax") {// use additional iterations to anneal to 0
	iters = 3*iters;
	temp = 1;
	for (Count i = 1; i <=temp_incr*2; i++) {
	  temp *= 1.2;
	  temperatures.push_back(temp);
	}
      }
      cout << "Raising temperature in " << temp_incr << " increments: " 
	   << temperatures << endl;
    }
    if (print_stats)
      state.print_stats_header(stats_os);

    //begin sampling loop
    Count temp_index = 0;
    for (Count i=0; i<iters; i++) {
      if ((i%10) == 0) cerr << ".";
      if (anneal && ((i%iter_incr) == 0)){
	temp = temperatures[temp_index++];
	cerr << "iter " << i << ": temp = " << temp << endl;
      }
      if (print_freq && 
	  ((i==100) || (i % print_freq == 0))) {
	cerr << i << " p_cont=" << state.p_cont() 
	     << " " << state.log_posterior() << endl;
	cout << "Before iteration " << i << 
	  ", with p_cont=" << state.p_cont() <<
	  ", log posterior = " << state.log_posterior() << ":" << endl;
	state.score_utterances(scoring);
	scoring.print_segmentation_summary();
	scoring.print_results();
	scoring.reset();
      }
      if (print_stats && (i % stats_freq == 0)) {
	stats_os << i << ",\t";
	state.print_stats(stats_os);
      }
      if (print_words && words_freq && (i % words_freq == 0)) {
	words_os << state << endl;
      }
      if (print_end && (i%words_freq == 0) &&
	  (!anneal ||
	   (temp == temperatures.back() && i%iter_incr >= iter_incr/2))) {
	stats_os << i << ",\t";
	state.print_stats(stats_os);
	words_os << state << endl;
      }
      state.sample(temp);
    } //end of sampling loop

    if (eval == "lmax")
	state.sample(10000); //like doing a local max instead of sample.
    //cout << state.get_lexicon() << endl;

    //print final stats
    if (print_stats || print_end) {
      stats_os << iters << ",\t";
      state.print_stats(stats_os);
    }
    if (print_words || print_end) {
      words_os << state;
    }

    state.score_utterances(scoring);
    if (verbose_level == 1 || verbose_level == 3) {
      cout << "State: " << endl;
      cout << state;
      // don't remove this line without making sure it isn't
      // assumed by scripts to be in the output.
      cout << "_nstrings=xxx" << endl << endl;
    }
    if (arguments.isset('l')) {
      scoring.print_reference_lexicon();
    }
    else {
    cout << endl;
    cerr << iters << " iterations" << endl;
    if (verbose_level == 2 || verbose_level == 3) {
      scoring.print_segmented_lexicon();
      cout << endl;
    }
    if (verbose_level > -1) 
      scoring.print_final_results();
    cout << "p_cont=" << state.p_cont() 
	 << ", log prob = " << state.log_posterior() << endl;
    }
//     cout << "Generating utterances:" << endl;
//     for (Count j = 0; j < 1000; j++) {
//     state.generate();
//     }
    delete data;
  }
  catch (FileError& e) {
    cout << "Error: couldn't open input file " << filename << endl;
  }
  catch (UtteranceLength& e) {
    e.print();
  }
}

